### PR TITLE
Implement einsum for (at least some combinations) of arrays with ragged dimensions

### PR DESCRIPTION
**Pull request by axch**

### CODE REVIEW

File: jax/_src/interpreters/batching.py [hunk 1]
```diff
@@ -134,6 +134,10 @@ def move_axis(ax):
     new_ragged_axes = [(move_axis(ax), sizes) for ax, sizes in self.ragged_axes]
     return RaggedAxis(dst, new_ragged_axes)
 
+  def transpose_ragged_axes(self, perm):
```
Comment (mattjj) [inline jax/_src/interpreters/batching.py:137]: IMO module-level functions are nicer than methods (for this and `move_stacked_axis`) but nbd!

File: jax/_src/interpreters/batching.py [hunk 2]
```diff
@@ -143,6 +147,22 @@ def make_batch_axis(
   else:
     return canonicalize_axis(stacked_axis, ndim)
 
+def bdim_as_shape(bdim, data_shape):
```
Comment (mattjj) [inline jax/_src/interpreters/batching.py:150]: Type annotations? Does Python even work without type annotations on function definitions!?

**Review by mattjj — APPROVED** [review]
Nice! Optional suggestions.

File: jax/_src/interpreters/batching.py (hunk 1 above)
Comment (axch) [reply jax/_src/interpreters/batching.py:137]: Will move in a future PR.

File: jax/_src/interpreters/batching.py (hunk 2 above)
Comment (axch) [reply jax/_src/interpreters/batching.py:150]: Per contra in Haskell, the type annotations are optional but the program is always type-correct anyway.

Added.

**Pull request merged**